egs/wsj/s5/steps/cleanup/internal/tf_idf.py

```diff
@@ -0,0 +1,398 @@
+                return 1 + math.log(self.raw_counts[(term, doc)])
+            return 0
+        if weighting_scheme == "normalized":
+            return (normalization_factor + (1 - normalization_factor) *
+                    self.raw_counts.get((term, doc), 0) /
+                    (1.0 + self.max_counts_for_term.get(term, 0)))
+        raise KeyError("Unknown tf-weighting-scheme {0}".format(
+                            weighting_scheme))
+
+    def accumulate(self, doc, text, ngram_order):
+        """Accumulate raw stats from a document for upto the specified
+        ngram-order."""
+        for n in range(1, ngram_order + 1):
+            for i in range(len(text)):
+                term = tuple(text[i:(i+n)])
+                self.raw_counts.setdefault((term, doc), 0)
+                self.raw_counts[(term, doc)] += 1
+
+    def compute_term_stats(self, idf_stats=None):
+        """Compute the maximum counts for each term over all the documents
+        based on the stored raw counts."""
+        if len(self.raw_counts) == 0:
+            raise kaldi_exceptions.EmptyObjectError(
+                "No (term, doc) found in tf-stats.")
+        for tup, counts in self.raw_counts.iteritems():
+            term = tup[0]
+
+            if counts > self.max_counts_for_term.get(term, 0):
+                self.max_counts_for_term[term] = counts
+
+            if idf_stats is not None:
+                idf_stats.accumulate(term)
+
+    def __str__(self):
+        """Returns a string with all the stats in the following format:
+        <n-gram order> <term-1> <term-2> ... <term-n> <document-id> <counts>
+        """
+        lines = []
+        for tup, counts in self.raw_counts.iteritems():
+            term, doc = tup
+            lines.append("{order} {term} {doc} {counts}".format(
+                            order=len(term), term=" ".join(term),
+                            doc=doc, counts=counts))
+        return "\n".join(lines)
+
+    def read(self, file_handle, ngram_order=None, idf_stats=None):
+        """Reads the TF stats stored in a file in the following format:
+        <ngram-order> <term-1> <term-2> ... <term-n> <document-id> <counts>
+
+        If idf_stats is provided then idf_stats is accumulated simultaneously.
+        """
+        for line in file_handle:
+            parts = line.strip().split()
+            order = parts[0]
+            assert(len(parts) - 3 == order)
+            if ngram_order is not None and order > ngram_order:
+                continue
+            term = tuple(parts[1:(order+1)])
+            doc = parts[-2]
+            counts = float(parts[-1])
+
+            self.raw_counts[(term, doc)] = counts
+
+            if counts > self.max_counts_for_term.get(term, 0):
+                self.max_counts_for_term[term] = counts
+
+            if idf_stats is not None:
+                idf_stats.accumulate(term)
+
+        if len(self.raw_counts) == 0:
+            raise kaldi_exceptions.EmptyObjectError(
```

Comment:
I think I'd prefer if you remove these "special" exception types unless you can demonstrate how they are actually useful (i.e. would need to be caught separately from other exceptions).